client/client-multi/client-c/src/kaa/kaa_notification_manager.c

```diff
@@ -0,0 +1,1001 @@
+        }
+        current_listener_node = kaa_list_next(current_listener_node);
+    }
+    return KAA_ERR_NONE;
+}
+
+static kaa_error_t kaa_notify_mandatory_notification_subscribers(kaa_notification_manager_t *self, uint64_t *topic_id, kaa_notification_t *notification)
+{
+    KAA_RETURN_IF_NIL2(self, notification, KAA_ERR_BADPARAM);
+    kaa_list_t *current_listener_node = self->mandatory_listeners;
+    KAA_RETURN_IF_NIL(current_listener_node,KAA_ERR_BAD_STATE);
+    kaa_notification_listener_wrapper_t *wrapper = NULL;
+    while (current_listener_node) {
+         wrapper = (kaa_notification_listener_wrapper_t *) kaa_list_get_data(current_listener_node);
+        if (wrapper->listener.callback) {
+            wrapper->listener.callback(wrapper->listener.context, topic_id, notification);
+        } else {
+            KAA_LOG_ERROR(self->logger, KAA_ERR_BADDATA, "Get NULL, expected callback, listener's id = %d.", wrapper->id);
+        }
+        current_listener_node = kaa_list_next(current_listener_node);
+    }
+    return KAA_ERR_NONE;
+}
+
+static kaa_error_t kaa_notify_optional_notification_subscribers(kaa_notification_manager_t *self, uint64_t *topic_id, const kaa_notification_t *notification)
+{
+    KAA_RETURN_IF_NIL3(self, topic_id, notification, KAA_ERR_BADPARAM);
+    kaa_list_t *opt_listeners_node = kaa_list_find_next(self->optional_listeners, &kaa_find_optional_notification_listener_by_id, topic_id);
+
+    if (!opt_listeners_node) {
+        KAA_LOG_WARN(self->logger, KAA_ERR_NOT_FOUND, "No listeners were notified, because nobody's subscribed on this topic (topic id = %d).", *topic_id);
+        return KAA_ERR_NOT_FOUND;
+    }
+    KAA_RETURN_IF_NIL(opt_listeners_node, KAA_ERR_BADPARAM);
+    kaa_optional_notification_listeners_wrapper_t* optional_wrapper = (kaa_optional_notification_listeners_wrapper_t *) kaa_list_get_data(opt_listeners_node);
+    kaa_list_t *optional_listener_node = optional_wrapper->listeners;
+    kaa_notification_listener_wrapper_t *wrapper = NULL;
+    while (optional_listener_node) {
+        wrapper = (kaa_notification_listener_wrapper_t *) kaa_list_get_data(optional_listener_node);
+        if (wrapper->listener.callback) {
+            wrapper->listener.callback(wrapper->listener.context, topic_id, notification);
+        } else {
+            KAA_LOG_ERROR(self->logger, KAA_ERR_BADDATA, "Get NULL, expected callback, listener's id = %d.", wrapper->id);
+        }
+        optional_listener_node = kaa_list_next(optional_listener_node);
+    }
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t kaa_subscribe_to_topic(kaa_notification_manager_t *self, uint64_t *topic_id, bool force_sync)
+{
+    KAA_RETURN_IF_NIL2(self, topic_id, KAA_ERR_BADPARAM);
+    kaa_topic_t* topic = NULL;
+    kaa_error_t err = kaa_find_topic(self, &topic, topic_id);
+    if (err) {
+        KAA_LOG_WARN(self->logger, KAA_ERR_BADPARAM, "Failed to subscribe to the topic with id = %d.", *topic_id);
```

Comment:
Log the err value instead KAA_ERR_BADPARAM. Please, review all this source file.